# Projet 2 : Analysez des données de systèmes éducatifs

Pour la pré-analyse, pourrais-tu :
* Valider la qualité de ce jeu de données (comporte-t-il beaucoup de données manquantes, dupliquées ?)
* Décrire les informations contenues dans le jeu de données (nombre de colonnes ? nombre de lignes ?)
* Sélectionner les informations qui semblent pertinentes pour répondre à la problématique (quelles sont les colonnes contenant des informations qui peuvent être utiles pour répondre à la problématique de l’entreprise ?)
* Déterminer des ordres de grandeurs des indicateurs statistiques classiques pour les différentes zones géographiques et pays du monde (moyenne/médiane/écart-type par pays et par continent ou bloc géographique)


## Importation des données

In [79]:
import pandas as pd

In [80]:
data_raw = pd.read_csv("EdStatsData.csv")
print("Le jeu de données contient {} lignes et {} colonnes.".format(data_raw.shape[0],data_raw.shape[1]))

Le jeu de données contient 886930 lignes et 70 colonnes.


In [31]:
data_raw.head()

Country Name Country Code  \
0   Arab World          ARB   
1   Arab World          ARB   
2   Arab World          ARB   
3   Arab World          ARB   
4   Arab World          ARB   

                                      Indicator Name  Indicator Code  \
0  Adjusted net enrolment rate, lower secondary, ...      UIS.NERA.2   
1  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.F   
2  Adjusted net enrolment rate, lower secondary, ...  UIS.NERA.2.GPI   
3  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.M   
4  Adjusted net enrolment rate, primary, both sex...     SE.PRM.TENR   

        1970       1971       1972       1973       1974      1975  ...  2060  \
0        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
1        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
2        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
3        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
4  54.822121  54.894138  56.209438  57.267109  57.991138  59.36554  ...   NaN   

   2065  2070  2075  2080  2085  2090  2095  2100  Unnamed: 69  
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
2   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
4   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  

[5 rows x 70 columns]

In [73]:
data_raw.tail()

Country Name Country Code  \
886925     Zimbabwe          ZWE   
886926     Zimbabwe          ZWE   
886927     Zimbabwe          ZWE   
886928     Zimbabwe          ZWE   
886929     Zimbabwe          ZWE   

                                           Indicator Name  \
886925  Youth illiterate population, 15-24 years, male...   
886926  Youth literacy rate, population 15-24 years, b...   
886927  Youth literacy rate, population 15-24 years, f...   
886928  Youth literacy rate, population 15-24 years, g...   
886929  Youth literacy rate, population 15-24 years, m...   

              Indicator Code  1970  1971  1972  1973  1974  1975  ...  2060  \
886925      UIS.LP.AG15T24.M   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
886926     SE.ADT.1524.LT.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
886927  SE.ADT.1524.LT.FE.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
886928  SE.ADT.1524.LT.FM.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
886929  SE.ADT.1524.LT.MA.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   

        2065  2070  2075  2080  2085  2090  2095  2100  Unnamed: 69  
886925   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
886926   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
886927   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
886928   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
886929   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  

[5 rows x 70 columns]

Nous observons, à partir des premières et dernières lignes :
* Le nom du pays (ou de la région) est indiqué dans la première colonne (index 0)
* le code du pays dans la deuxième colonne (index 1)
* le nom de l'indicateur est indiqué dans la troisième colonne (index 2)
* le code de l'indicateur est indiqué dans la quatrième colonne (index 3)
* les colonnes suivantes donnent les valeurs de l'indicateur depuis 1970 (index 4) à 2090.

Les données sont donc multi-indexé sur le pays ("Country code") et l'indicateur ("Indicator code").

In [178]:
liste_pays = list(data_raw["Country Name"].unique())
print("Nombre de pays et de régions dans le jeu de données :", len(liste_pays))
liste_pays

Nombre de pays et de régions dans le jeu de données : 242


['Arab World',
 'East Asia & Pacific',
 'East Asia & Pacific (excluding high income)',
 'Euro area',
 'Europe & Central Asia',
 'Europe & Central Asia (excluding high income)',
 'European Union',
 'Heavily indebted poor countries (HIPC)',
 'High income',
 'Latin America & Caribbean',
 'Latin America & Caribbean (excluding high income)',
 'Least developed countries: UN classification',
 'Low & middle income',
 'Low income',
 'Lower middle income',
 'Middle East & North Africa',
 'Middle East & North Africa (excluding high income)',
 'Middle income',
 'North America',
 'OECD members',
 'South Asia',
 'Sub-Saharan Africa',
 'Sub-Saharan Africa (excluding high income)',
 'Upper middle income',
 'World',
 'Afghanistan',
 'Albania',
 'Algeria',
 'American Samoa',
 'Andorra',
 'Angola',
 'Antigua and Barbuda',
 'Argentina',
 'Armenia',
 'Aruba',
 'Australia',
 'Austria',
 'Azerbaijan',
 'Bahamas, The',
 'Bahrain',
 'Bangladesh',
 'Barbados',
 'Belarus',
 'Belgium',
 'Belize',
 'Benin',
 'Berm

Nous constatons qu'il y a un total de 242 pays et régions.

Les régions sont situées au début de la base de données (par ordre alphabétique, de 'Arab World' à 'World').

Ensuite viennent les pays, également classés par ordre alphabétique (de 'Afghanistan' à 'Zimbabwe').

Nous allons séparer la liste des pays et régions.

In [179]:
liste_regions = []
while 'World' in liste_pays:
    liste_regions.append(liste_pays.pop(0))
print("Liste des régions : de {} à {}".format(liste_regions[0], liste_regions[-1]))
print("Liste des pays : de {} à {}".format(liste_pays[0], liste_pays[-1]))

Liste des régions : de Arab World à World
Liste des pays : de Afghanistan à Zimbabwe


In [82]:
liste_indicateurs = data_raw["Indicator Name"].unique()
print("Nombre d'indicateurs dans le jeu de données :", len(liste_indicateurs))

Nombre d'indicateurs dans le jeu de données : 3665


Il y a un très grand nombre d'indicateurs. Il ne sera bien évidemment pas possible de tous les exploiter, et nous devrons sélectionner les plus pertinents pour notre analyse.

## Identification et traitement des doublons

In [47]:
# Les valeurs dupliquées correspondent aux lignes qui ont le même pays et le même indicateur.
duplicates = data_raw.duplicated(subset=["Country Code","Indicator Code"], keep='first')
duplicates.unique()

array([False])

Nous constatons qu'il n'y a pas de données dupliquées (même code pays et même code indicateur).

## Sélection des colonnes pertinentes
Nous décidons de ne garder les données que pour les années 2012 à 2030. Cela nous informera suffisamment sur l'état actuel et les perspectives à moyen terme.

In [85]:
data = pd.concat([data_raw.loc[:,"Country Name":"Indicator Code"],data_raw.loc[:,"2012":"2030"]], axis=1)
# Nous affichons les premières lignes pour vérifier la structure des données
data.head()

Country Name Country Code  \
0   Arab World          ARB   
1   Arab World          ARB   
2   Arab World          ARB   
3   Arab World          ARB   
4   Arab World          ARB   

                                      Indicator Name  Indicator Code  \
0  Adjusted net enrolment rate, lower secondary, ...      UIS.NERA.2   
1  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.F   
2  Adjusted net enrolment rate, lower secondary, ...  UIS.NERA.2.GPI   
3  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.M   
4  Adjusted net enrolment rate, primary, both sex...     SE.PRM.TENR   

        2012      2013       2014  2015  2016  2017  2020  2025  2030  
0        NaN       NaN        NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1        NaN       NaN        NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2        NaN       NaN        NaN   NaN   NaN   NaN   NaN   NaN   NaN  
3        NaN       NaN        NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4  86.101669  85.51194  85.320152   NaN   NaN   NaN   NaN   NaN   NaN

## Valeurs manquantes

In [86]:
# Résumé statistique
data.describe() 

2012          2013          2014          2015          2016  \
count  1.472640e+05  1.375090e+05  1.137890e+05  1.310580e+05  1.646000e+04   
mean   2.176323e+10  2.425332e+10  2.678449e+10  2.353720e+10  1.934411e+11   
std    8.833955e+11  9.513016e+11  1.023180e+12  9.732465e+11  2.839188e+12   
min   -6.049930e+05 -6.157480e+05 -8.900000e+01 -2.467847e+00 -1.797539e+00   
25%    1.100000e+01  1.306113e+01  1.614639e+01  4.100000e-01  1.826761e+01   
50%    9.759012e+01  1.000000e+02  1.000000e+02  5.235000e+01  6.264000e+03   
75%    1.038168e+05  1.426480e+05  1.636440e+05  6.153575e+04  5.939590e+05   
max    9.999473e+13  1.054580e+14  1.108060e+14  1.156198e+14  1.206030e+14   

             2017          2020          2025          2030  
count  143.000000  5.143600e+04  5.143600e+04  5.143600e+04  
mean     2.160839  5.780697e+02  6.054774e+02  6.315916e+02  
std      0.810584  1.475072e+04  1.577305e+04  1.678556e+04  
min      1.000000 -1.870000e+00 -2.000000e+00 -2.110000e+00  
25%      2.000000  6.000000e-02  6.000000e-02  6.000000e-02  
50%      2.000000  2.300000e-01  2.300000e-01  2.300000e-01  
75%      3.000000  6.660000e+00  6.900000e+00  7.062500e+00  
max      4.000000  1.599479e+06  1.781898e+06  1.967258e+06

Nous voyons que de très nombreuses valeurs sont manquantes pour chaque année. En effet, il y a un total de 886930 (8,86e+5) lignes.

En particulier, nous constatons que seulement 143 valeurs sont renseignées pour l'année 2017. Nous décidons donc de suprimer cette colonne.

In [87]:
data = data.drop("2017", axis=1)

In [72]:
data.head()

Country Name Country Code  \
0   Arab World          ARB   
1   Arab World          ARB   
2   Arab World          ARB   
3   Arab World          ARB   
4   Arab World          ARB   

                                      Indicator Name  Indicator Code  \
0  Adjusted net enrolment rate, lower secondary, ...      UIS.NERA.2   
1  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.F   
2  Adjusted net enrolment rate, lower secondary, ...  UIS.NERA.2.GPI   
3  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.M   
4  Adjusted net enrolment rate, primary, both sex...     SE.PRM.TENR   

        2012      2013       2014  2015  2016  2020  2025  2030  
0        NaN       NaN        NaN   NaN   NaN   NaN   NaN   NaN  
1        NaN       NaN        NaN   NaN   NaN   NaN   NaN   NaN  
2        NaN       NaN        NaN   NaN   NaN   NaN   NaN   NaN  
3        NaN       NaN        NaN   NaN   NaN   NaN   NaN   NaN  
4  86.101669  85.51194  85.320152   NaN   NaN   NaN   NaN   NaN

In [144]:
# Nous allons séléctionner les indicateurs qui ont les meilleurs taux de remplissage.
# Pour cela, nous allons agréger les données par indicateur et calculer pour chaque indicateur le taux de remplissage.

gb_indicateurs = data.groupby('Indicator Code')
print(gb_indicateurs.sum()[::200])

                                   2012          2013          2014  \
Indicator Code                                                        
BAR.NOED.1519.FE.ZS        0.000000e+00  0.000000e+00  0.000000e+00   
BAR.SEC.CMPT.5559.FE.ZS    0.000000e+00  0.000000e+00  0.000000e+00   
HH.DHS.OOS.1               2.891000e+02  1.512699e+02  1.611737e+02   
HH.MICS.YRS.15UP.GIN       4.849222e+01  0.000000e+00  2.683296e+02   
LO.EGRA.LSTN.0.SID.2GRD    0.000000e+00  0.000000e+00  0.000000e+00   
LO.EGRA.READ.KNY.ADV.6GRD  0.000000e+00  0.000000e+00  0.000000e+00   
LO.PASEC.MAT.6.P10         0.000000e+00  0.000000e+00  3.926500e+03   
LO.PIRLS.REA.P25           0.000000e+00  0.000000e+00  0.000000e+00   
LO.TIMSS.MAT8.MA           0.000000e+00  0.000000e+00  0.000000e+00   
PRJ.ATT.25UP.3.FE          0.000000e+00  0.000000e+00  0.000000e+00   
SABER.EMIS.GOAL1.LVL6      0.000000e+00  0.000000e+00  0.000000e+00   
SABER.WORK.GOAL1.LVL3      2.100000e+01  2.100000e+01  2.100000e+01   
SL.UEM

Nous constatons que certains indicateurs ne sont pas remplis du tout, par aucun pays et aucune année.
Nous allons déjà supprimer ces indicateurs vides.

In [161]:
# Pour chaque indicateur, nous allons faire le décompte des valeurs non-nulles

col_list = list(gb_indicateurs_count)
col_list.remove('Country Name')
col_list.remove('Country Code')
col_list.remove('Indicator Name')
col_list

# Séparer en valeurs "passées" et valeurs "futur"

gb_indicateurs_count = gb_indicateurs.count()
gb_indicateurs_count['Total'] = gb_indicateurs_count[col_list].sum(axis=1)
gb_indicateurs_count[::200] # affiche une valeur sur 200

Country Name  Country Code  Indicator Name  2012  \
Indicator Code                                                                
BAR.NOED.1519.FE.ZS                 242           242             242     0   
BAR.SEC.CMPT.5559.FE.ZS             242           242             242     0   
HH.DHS.OOS.1                        242           242             242    14   
HH.MICS.YRS.15UP.GIN                242           242             242     1   
LO.EGRA.LSTN.0.SID.2GRD             242           242             242     0   
LO.EGRA.READ.KNY.ADV.6GRD           242           242             242     0   
LO.PASEC.MAT.6.P10                  242           242             242     0   
LO.PIRLS.REA.P25                    242           242             242     0   
LO.TIMSS.MAT8.MA                    242           242             242     0   
PRJ.ATT.25UP.3.FE                   242           242             242     0   
SABER.EMIS.GOAL1.LVL6               242           242             242     0   
SABER.WORK.GOAL1.LVL3               242           242             242     7   
SL.UEM.TOTL.ZS                      242           242             242   211   
SP.PRM.GRAD.TO                      242           242             242   216   
UIS.E.2.GPV.F                       242           242             242   193   
UIS.G.7.T                           242           242             242    10   
UIS.NART.2.RUR                      242           242             242     7   
UIS.REPR.2.GPV.G4                   242           242             242    41   
UIS.XPUBP.2                         242           242             242    65   

                           2013  2014  2015  2016  2020  2025  2030  Total  
Indicator Code                                                              
BAR.NOED.1519.FE.ZS           0     0     0     0     0     0     0      0  
BAR.SEC.CMPT.5559.FE.ZS       0     0     0     0     0     0     0      0  
HH.DHS.OOS.1                  7     8     6     0     0     0     0     35  
HH.MICS.YRS.15UP.GIN          0     7     0     0     0     0     0      8  
LO.EGRA.LSTN.0.SID.2GRD       0     0     0     0     0     0     0      0  
LO.EGRA.READ.KNY.ADV.6GRD     0     0     0     0     0     0     0      0  
LO.PASEC.MAT.6.P10            0    10     0     0     0     0     0     10  
LO.PIRLS.REA.P25              0     0     0    45     0     0     0     45  
LO.TIMSS.MAT8.MA              0     0    35     0     0     0     0     35  
PRJ.ATT.25UP.3.FE             0     0   167     0   167   167   167    668  
SABER.EMIS.GOAL1.LVL6         0     0     4     2     0     0     0      6  
SABER.WORK.GOAL1.LVL3         9    10     1     0     0     0     0     27  
SL.UEM.TOTL.ZS              211   208   208   208     0     0     0   1046  
SP.PRM.GRAD.TO              216   216   196   154     0     0     0    998  
UIS.E.2.GPV.F               182   161     8     0     0     0     0    544  
UIS.G.7.T                    75    74     3     0     0     0     0    162  
UIS.NART.2.RUR                8     2     1     0     0     0     0     18  
UIS.REPR.2.GPV.G4            38     3     0     0     0     0     0     82  
UIS.XPUBP.2                  52    26     0     0     0     0     0    143

Nous constatons que certains indicateurs ne sont pas remplis du tout, ou très peu. Comme il y a un total de 8 colonnes pour les années (2012 à 2030), et 242 pays ou régions, un indicateur totalement rempli devrait comporter 1936 valeurs. En ce limitant aux valeurs passées, années 2012 à 2016, nous devrions avoir 1210 valeurs.

In [162]:
# Tri des indicateurs par nombre décroissant de réponses
data = gb_indicateurs_count.sort_values("Total",ascending=False)
data[::200] # affiche une valeur sur 200

Country Name  Country Code  Indicator Name  \
Indicator Code                                                               
SP.POP.GROW                              242           242             242   
SP.POP.1116.MA.UN                        242           242             242   
PRJ.ATT.2024.NED.FE                      242           242             242   
SP.POP.AG17.FE.UN                        242           242             242   
UIS.E.0.T                                242           242             242   
UIS.FOSEP.56.F400.M                      242           242             242   
UIS.XGDP.1.FSGOV.FDINSTADM.FFD           242           242             242   
LO.PISA.REA.6.FE                         242           242             242   
UIS.AFR.SCHBSP.1.PU.WOTOIL               242           242             242   
UIS.AFR.FNTP.1                           242           242             242   
UIS.AFR.SCHBSP.2.PU.WNIPOWAT             242           242             242   
UIS.E.2.GPV.G6.F                         242           242             242   
SABER.PRVT.GOAL4.LVL4                    242           242             242   
SABER.SCH.FNNC.GOAL4                     242           242             242   
HH.MICS.TR.12.U                          242           242             242   
BAR.PRM.SCHL.4044                        242           242             242   
LO.TIMSS.MAT8.INT.FE                     242           242             242   
LO.EGRA.CWPM.ZERO.SID.2GRD               242           242             242   
BAR.SEC.SCHL.6569.FE                     242           242             242   

                                2012  2013  2014  2015  2016  2020  2025  \
Indicator Code                                                             
SP.POP.GROW                      239   239   232   232   232     0     0   
SP.POP.1116.MA.UN                181   181   181   181     0     0     0   
PRJ.ATT.2024.NED.FE                0     0     0   167     0   167   167   
SP.POP.AG17.FE.UN                181   181     0   181     0     0     0   
UIS.E.0.T                        127   128   110     7     0     0     0   
UIS.FOSEP.56.F400.M               90    90    77     4     0     0     0   
UIS.XGDP.1.FSGOV.FDINSTADM.FFD    81    69    35     0     0     0     0   
LO.PISA.REA.6.FE                  64     0     0    71     0     0     0   
UIS.AFR.SCHBSP.1.PU.WOTOIL        24    19    11     1     0     0     0   
UIS.AFR.FNTP.1                    15    16     2     0     0     0     0   
UIS.AFR.SCHBSP.2.PU.WNIPOWAT       6     8     2     2     0     0     0   
UIS.E.2.GPV.G6.F                   2     5     9     0     0     0     0   
SABER.PRVT.GOAL4.LVL4              0     0     0     3     6     0     0   
SABER.SCH.FNNC.GOAL4               3     1     0     0     0     0     0   
HH.MICS.TR.12.U                    0     0     1     0     0     0     0   
BAR.PRM.SCHL.4044                  0     0     0     0     0     0     0   
LO.TIMSS.MAT8.INT.FE               0     0     0     0     0     0     0   
LO.EGRA.CWPM.ZERO.SID.2GRD         0     0     0     0     0     0     0   
BAR.SEC.SCHL.6569.FE               0     0     0     0     0     0     0   

                                2030  Total  
Indicator Code                               
SP.POP.GROW                        0   1174  
SP.POP.1116.MA.UN                  0    724  
PRJ.ATT.2024.NED.FE              167    668  
SP.POP.AG17.FE.UN                  0    543  
UIS.E.0.T                          0    372  
UIS.FOSEP.56.F400.M                0    261  
UIS.XGDP.1.FSGOV.FDINSTADM.FFD     0    185  
LO.PISA.REA.6.FE                   0    135  
UIS.AFR.SCHBSP.1.PU.WOTOIL         0     55  
UIS.AFR.FNTP.1                     0     33  
UIS.AFR.SCHBSP.2.PU.WNIPOWAT       0     18  
UIS.E.2.GPV.G6.F                   0     16  
SABER.PRVT.GOAL4.LVL4              0      9  
SABER.SCH.FNNC.GOAL4               0      4  
HH.MICS.TR.12.U                    0      1  
BAR.PRM.SCHL.4

In [154]:
# Nous allons construire un histogramme présentant le nombre de valeurs par indicateur
